Adding Slotfill extension components for remote payments

joelclimbsthings opened the pull request:
Fixes #6853 

This is a starting point for a potential Slotfill solution for remote payments. It's one potential approach, and there is freedom here to be fairly strategic with how we place/configure Slotfills. I've tried to do so in a way that maximized capability while keeping it fairly simple to use. 

I've also created a `<SettingsForm />` component that accepts a variety of props including `fields`. The fields should match the format as those stored in the settings API (or an array of the values in that object). Right now I'm just filtering these to only those that "appear" in the remote config, but this will be altered to tap into @joshuatf 's revisions to the API that will determine essential fields. Right now it only supports text, password, checkbox and select type fields. It's fairly immature so far, but works well for this use case and is ripe for further extension. Open for any feedback on what we want to include/change on this in the context of this PR. 
```js
<SettingsForm
	fields={ fields }
	isBusy={ isOptionsRequesting }
	onSubmit={ updateSettings }
	onButtonClick={ () => recordConnectStartEvent( key ) }
	buttonLabel={ __( 'Proceed', 'woocommerce-admin' ) }
	validate={ validate }
	{ ...props }
/>
```

Details on the Slotfills:
#### `<WooRemotePayment />`
This is a Slot over the entire `<Stepper />` that can ultimately allow for customization of the entire setup/configuration process, including adding steps, etc. 
```js
<WooRemotePayment.Slot
	defaultStepper={ DefaultStepper }
	defaultInstallStep={ installStep }
	defaultConnectStep={ connectStep }
	id={ key }
/>
```

It also passes some helpful props:
**defaultStepper**: The default, pre-configured stepper component that can be customized by overriding specific props.
**defaultInstallStep**: The default installation step, in case they just want to add steps, and plug in the defaults.
**defaultConnectStep**: Same as above, for the connect step.

#### `<WooRemotePaymentSettings />`
We're using the new `<Settings />` component with `<PaymentConnect />`, and this slot will be used as an alternative. 
```js
<WooRemotePaymentSettings.Slot
	defaultSettings={ DefaultSettings }
	defaultUpdate={ updateSettings }
	markConfigured={ () => markConfigured( key ) }
	id={ key }
/>
```
Helpful props:
**defaultSettings**: The default settings component, with can have individual props overridden.
**defaultUpdate**: The default update function passed to `<Settings />`.
**markConfigured**: The callback to mark the method as configured, but no need to pass the method key.


### Screenshots

![image](https://user-images.githubusercontent.com/444632/117180663-441fed80-ad89-11eb-9592-5e2ad4d0c8da.png)

### Detailed test instructions:

- Checkout branch
- Enable remote payments feature flag
- You can checkout [my fork of the remote payment tester](https://github.com/joelclimbsthings/remote-payments-tester) with added front-end tooling (I can merge this into the original if we want).
- Activate the remote payment tester, and run `npm install` and `npm start`
- Navigate to the payment task and click on one of the sample configs (payfast, stripe). 
- You should see the fields as pictured in the screenshot. These are built from the settings API, but limited by the sample configuration provided by the tester plugin (again, this will change a bit after enhancements to the API).
- Notice the custom headings. These are inserted via Slotfill, and the `update` function is overridden as well to just do a `console.info` from the tester plugin.
- Poke around with the demo Slotfills in the `src/index.js` of the tester plugin, and temporarily remove them to ensure that the default functionality steps in as it should.

joelclimbsthings committed:
Working slotfill for payment method stepper

joelclimbsthings committed:
Laying the groundwork for general Settings component

joelclimbsthings committed:
Settings component working for all basic types

joelclimbsthings committed:
Moving API logic to Settings component, adding Settings SlotFill

joelclimbsthings committed:
Tweaks to loading and error states for Settings component

joelclimbsthings committed:
Using Text component

joelclimbsthings committed:
Moving validation outside of Settings for now, removing required props

joelclimbsthings committed:
Moving api call out of Settings, simplifying some logic

joelclimbsthings committed:
Renaming props for clarity

psealock commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +137,45 @@ export const PaymentConnect = ( {
+					defaultSettings={ DefaultSettings }
+					defaultUpdate={ updateSettings }
+					markConfigured={ () => markConfigured( key ) }
+					id={ key }
```
You could also pass in `fields` here as well under `defaultFields` so the extension can decide if they want to use `DefaultSettings` or create whatever else they want.

psealock commented on packages/components/src/settings/index.js:
```diff
@@ -0,0 +1,84 @@
+} ) => {
+	// Support accepting fields in the format provided by the API (object), but transform to Array
+	const fields =
+		baseFields instanceof Array ? baseFields : Object.values( baseFields );
```
In a README, it would be good to document the expected shape of the data.

joelclimbsthings replied:
Yup, definitely on the agenda.

psealock commented on packages/components/src/settings/index.js:
```diff
@@ -0,0 +1,84 @@
+	select: SettingSelect,
+};
+
+export const Settings = ( {
```
This component could be called `SettingsForm`, what do you think?

joelclimbsthings replied:
Sound good to me 👍🏻

psealock reviewed (commented):
This is truly amazing @joelclimbsthings !

I left some comments on some small things, but this is testing really well.

What do you think of pulling out the `Settings` component to its own PR? Your choice, but it could be useful for getting better feedback from other members of Isotope who have lot of experience with this code and functionality. I suspect you'll get suggested to use TypeScript because its a new component and it accepts data of a certain shape, which is probably a good idea.

I must say, this is really powerful and places lots of option in the hands of extensions 😄 . Having said that, what is that use case for replacing or amending the `DefaultStepper`? Are we inviting ourselves to overly complex situations? It seems to me that we can solve the problem with just `WooRemotePaymentSettings`, but I may be missing something.

Its really cool to see this come to life!

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +137,45 @@ export const PaymentConnect = ( {
+	if ( state === 'error' ) {
+		return (
+			<Text>
+				{ __( 'There was an error loading the payment fields' ) }
```
Missing text domain.

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +137,45 @@ export const PaymentConnect = ( {
 			onSubmitCallback={ updateSettings }
+			onButtonClickCallback={ () => recordConnectStartEvent( key ) }
+			buttonLabel={ __( 'Proceed', 'woocommerce-admin' ) }
 			validate={ validate }
```
Just thinking out loud here: I wonder if validation will happen on the input level (individual settings) or form (settings group) level.  I generally prefer on the form level as it is now since it allows validation by comparing multiple fields, but just wondering how that ties into existing validation with WC Settings.

My guess is that we should keep things as you have them here since client-side validation will have to be freshly added anyway, but just putting this out there to think about.

joelclimbsthings replied:
Yes, bit of an open question for me as well. I'll keep what I have currently and we may revisit soon.

joshuatf commented on packages/components/src/settings/index.js:
```diff
@@ -0,0 +1,84 @@
+	SettingSelect,
+} from './types';
+
+const typeMap = {
```
Nice!  Wondering if we should have any type of error handling if a type is invalid or default to text.

joelclimbsthings replied:
Good call, I'll throw in `text` as default and a warning if it's an unsupported type.

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +137,45 @@ export const PaymentConnect = ( {
+		<Settings
+			fields={ fields }
+			isBusy={ isOptionsRequesting }
 			onSubmitCallback={ updateSettings }
```
I know this is `onSubmitCallback` in form, but I'm in favor of moving this to `onSubmit`.  Relevant - https://github.com/woocommerce/woocommerce-admin/pull/6708

joelclimbsthings replied:
Done ✅

joshuatf reviewed (commented):
Nice, Joel!  This is going to prove really useful not only for this project, but future settings work.  💯 

Left a few quick clean-up comments, but nothing major.  In favor of shipping this and continuing to iterate.

joelclimbsthings commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -2,58 +2,69 @@
 	const { updateOptions } = useDispatch( OPTIONS_STORE_NAME );
 	const { createNotice } = useDispatch( 'core/notices' );
+	const slot = useSlot( `woocommerce_remote_payment_settings_${ key }` );
+	const hasFills = Boolean( slot.fills && slot.fills.length );
```
This changes everything. 💥

joelclimbsthings commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +137,45 @@ export const PaymentConnect = ( {
-	return (
-		<Form
-			initialValues={ getInitialConfigValues() }
+	const DefaultSettings = ( { ...props } ) => (
```
Ah yes! Thanks 🤭

joelclimbsthings commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -2,58 +2,69 @@
+				console.error( e );
+				/* eslint-enable no-console */
+			} );
+	}, [] );
```
Yup, there are a few small changes I'll need to address after some pending PRs go through. Right now I'm keen to merge things and iterate instead of having several PRs pile up waiting for eachother.

joelclimbsthings committed:
Implementing feedback, including renaming Settings to SettingsForm.

joelclimbsthings committed:
Renaming callbacks in SettingsForm

joelclimbsthings committed:
Adding support for post_install_script in config

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -87,6 +101,29 @@ export const PaymentConnect = ( {
+		if ( fields ) {
+			return fields.reduce( ( errors, field ) => {
+				if ( ! values[ field.id ] ) {
+					// Matches any word that is capitalized aside from abrevitions like ID.
```
🆒

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentMethod.js:
```diff
@@ -24,7 +26,15 @@ export const PaymentMethod = ( {
+		key,
+		plugins,
+		title,
+		post_install_script: postInstallScript,
```
This is ok for now, but just noting that this will definitely need to be moved to the payment gateway once https://github.com/woocommerce/woocommerce-admin/pull/6918 and https://github.com/woocommerce/woocommerce-admin/pull/6919 are merged.

joelclimbsthings replied:
Yup, agreed. The only downside I see is that it will required too non-parallel requests, instead of one, but it does seem the better place for it.

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentMethod.js:
```diff
@@ -71,6 +81,12 @@ export const PaymentMethod = ( {
+
+						if ( postInstallScript ) {
+							const script = document.createElement( 'script' );
+							script.src = `${ window.location.origin }/${ postInstallScript }`;
```
I _think_ this is okay given that the plugin defines this.  But we should probably not use `origin` since the WP install path is not always at the root.  Instead we can allow `postInstallScript` to simply define the full URL.

@psealock can I get a sanity check on the safety here?

joelclimbsthings replied:
Right, I wondered about that very thing. I'll remove the origin and expect the full URL, although this will be obsolete soon anyways.

joshuatf commented on packages/components/src/settings-form/index.js:
```diff
@@ -0,0 +1,95 @@
+								return null;
+							}
+
+							const Control = typeMap[ field.type || 'default' ];
```
Sorry for changing up here, but this logic may need some tweaking.  There's a possibility that this is defined by doesn't match one of the types (e.g., `textarea`) in which case I think we shouldn't attempt to show the control at all if we don't have a matching component.

joelclimbsthings replied:
Did you see the block of code right above this @joshuatf ? That's its intention. If a non-supported type is provided it won't render it at all, and will print a warning to the console.

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +140,49 @@ export const PaymentConnect = ( {
+		<SettingsForm
+			fields={ fields }
+			isBusy={ isOptionsRequesting }
 			onSubmitCallback={ updateSettings }
```
I'm still seeing this as `onSubmitCallback`- maybe the changes weren't pushed?

joelclimbsthings replied:
Good catch, I updated the component but not it's usage 👍🏻

joshuatf commented on client/task-list/tasks/payments/RemotePayments/components/PaymentConnect.js:
```diff
@@ -103,39 +140,49 @@ export const PaymentConnect = ( {
+			fields={ fields }
+			isBusy={ isOptionsRequesting }
 			onSubmitCallback={ updateSettings }
+			onButtonClickCallback={ () => recordConnectStartEvent( key ) }
```
Where does this connect to?  I'm not seeing this prop used anywhere.

joelclimbsthings replied:
I'd forgotten to update this as well. It's actually the `onButtonClick` handler, fired when the submit button is clicked.

joshuatf reviewed (commented):
Thanks for those changes, Joel!  Left a couple more, but I think this is just about ready.

Hoping @psealock can weigh in on the script injection safety.

joelclimbsthings committed:
Adding basic story for <SettingsForm />

joelclimbsthings committed:
Adding validation to story

joelclimbsthings committed:
Adding readme for SettingsForm

joelclimbsthings committed:
Adding README's for slot components, and removing stories

joelclimbsthings committed:
Refactoring slotfill components, and providing readmes

joelclimbsthings committed:
Removing use of location.origin and adding todo comment

joelclimbsthings committed:
Adding changelog

joelclimbsthings commented:
Thanks for so much valuable feedback @psealock & @joshuatf . I've just added documentation, refactored some things, and made some changes based on your feedback. Also added the changelog, so _hopefully_ we're about ready to merge 🥳 . 

> I suspect you'll get suggested to use TypeScript because its a new component and it accepts data of a certain shape, which is probably a good idea.

This is a totally valid ask @psealock , although with all of the parallel related PRs we have now I'd rather not create another PR that's a dependency. How about we get this merged, and I'll plan to create another PR for the typescript transformation/etc, at which time we can get more feedback?

> It seems to me that we can solve the problem with just WooRemotePaymentSettings, but I may be missing something.

This is totally true for our known use-cases, although I do think it would provide some pretty powerful options in terms of adding additional steps/etc. My expectation is that the `WooRemoteSettings` slotfill will be the simpler option for most, but that `WooRemotePayment` provides some cool capability that might be useful.

All that said, my hope is that we can get this merged, pending resolution on the above and Josh's question regarding script injection safety.

joelclimbsthings committed:
Fixing activity panel tests

psealock reviewed (approved):
Changes look good @joelclimbsthings, approving but I think expecting the full URL from the postInstall script url is a must-have.

At this point, I think its good to unlock the next level of the challenge by getting this merged and refine the `Settings` component in a follow up as well as address any gotchas that may arise.

If we don't get this over the line in before the next release, it may be worthwhile to move the component to `@woocommerce/experimental` but we can decide that later.
